packages/pigeon/lib/ast.dart

```diff
@@ -177,6 +178,46 @@ class AstProxyApi extends Api {
         (ApiField field) => !field.isAttached,
       );
 
+  /// A list of AstProxyApis where each `extends` the API that follows it.
+  Iterable<AstProxyApi> get allSuperClasses => recursiveGetSuperClassApisChain(
+        this,
+      );
+
+  /// All ProxyApis this API `implements` and all the interfaces those APIs
+  /// `implements`.
+  Iterable<AstProxyApi> get apisOfInterfaces =>
+      recursiveFindAllInterfaceApis(this);
+
+  /// All methods inherited from interfaces and the interfaces of interfaces.
+  Iterable<Method> flutterMethodsFromInterfaces() sync* {
+    for (final AstProxyApi proxyApi in apisOfInterfaces) {
+      yield* proxyApi.methods;
+    }
+  }
+
+  /// A list of Flutter methods inherited from the ProxyApi that this ProxyApi
+  /// `extends`.
+  ///
+  /// This also recursively checks the ProxyApi that the super class `extends`
+  /// and so on.
+  ///
+  /// This also includes methods that super classes inherited from interfaces
+  /// with `implements`.
+  Iterable<Method> flutterMethodsFromSuperClasses() sync* {
+    for (final AstProxyApi proxyApi in allSuperClasses.toList().reversed) {
+      yield* proxyApi.flutterMethods;
+    }
+    if (superClass != null) {
+      final Set<AstProxyApi> interfaceApisFromSuperClasses =
+          recursiveFindAllInterfaceApis(
+        superClass!.associatedProxyApi!,
+      );
+      for (final AstProxyApi proxyApi in interfaceApisFromSuperClasses) {
+        yield* proxyApi.methods;
+      }
+    }
+  }
```

Comment:
I don't know if performance really matters here, but ideally these methods would only ever be ran once, and the data stored on the class.

Reply:
Since some of the fields of `AstProxyApi` that this method uses are mutable, I don't think we can run it only once. And they are mutable because of the step in `pigeon_lib.dart` where the `associatedProxyApi` is added to all the type declarations.

Reply:
Doesn't the code that associates the api's make copies? Maybe I'm looking in the wrong place.

Reply:
Unless I'm missing something, a new api is not created: https://github.com/flutter/packages/blob/main/packages/pigeon/lib/pigeon_lib.dart#L1297 and https://github.com/flutter/packages/blob/main/packages/pigeon/lib/pigeon_lib.dart#L1337.

Only `NamedType`s (e.g. fields, methods parameters, method returnType, constructor parameters, etc...) are copied. The original apis, methods, and constructors are retained.

But even so, as long as the fields are mutable, it could cause a bug in the future if a contributor uses the method. I do think we should probably restructure the AST with immutable classes in the future. Mutable data classes can end up being a foot-gun.

Reply:
Can you add a todo for this and tag me in it?